cgroups/src/common.rs

```diff
@@ -91,92 +94,79 @@ pub fn read_cgroup_file<P: AsRef<Path>>(path: P) -> Result<String> {
     fs::read_to_string(path).with_context(|| format!("failed to open {:?}", path))
 }
 
-pub fn get_supported_cgroup_fs() -> Result<Vec<Cgroup>> {
-    let cgroup_mount = Process::myself()?
-        .mountinfo()?
-        .into_iter()
-        .find(|m| m.fs_type == "cgroup");
-
-    let cgroup2_mount = Process::myself()?
-        .mountinfo()?
-        .into_iter()
-        .find(|m| m.fs_type == "cgroup2");
+/// Determines the cgroup setup of the system. Systems typically have one of
+/// three setups:
+/// - Unified: Pure cgroup v2 system.
+/// - Legacy: Pure cgroup v1 system.
+/// - Hybrid: Hybrid is basically a cgroup v1 system, except for
+///   an additional unified hierarchy which doesn't have any
+///   controllers attached. Resource control can purely be achieved
+///   through the cgroup v1 hierarchy, not through the cgroup v2 hierarchy.
+pub fn get_cgroup_setup() -> Result<CgroupSetup> {
+    let default_root = Path::new(DEFAULT_CGROUP_ROOT);
+    match default_root.exists() {
+        true => {
+            // If the filesystem is of type cgroup2, the system is in unified mode.
+            // If the filesystem is tmpfs instead the system is either in legacy or
+            // hybrid mode. If a cgroup2 filesystem has been mounted under the "unified"
+            // folder we are in hybrid mode, otherwise we are in legacy mode.
+            let stat = statfs(default_root).with_context(|| {
+                format!(
+                    "failed to stat default cgroup root {}",
+                    &default_root.display()
+                )
+            })?;
+            if stat.filesystem_type() == CGROUP2_SUPER_MAGIC {
+                return Ok(CgroupSetup::Unified);
+            }
 
-    let mut cgroups = vec![];
-    if cgroup_mount.is_some() {
-        cgroups.push(Cgroup::V1);
-    }
+            if stat.filesystem_type() == TMPFS_MAGIC {
+                let unified = Path::new("/sys/fs/cgroup/unified");
+                if Path::new(unified).exists() {
+                    let stat = statfs(unified)
+                        .with_context(|| format!("failed to stat {}", unified.display()))?;
+                    if stat.filesystem_type() == CGROUP2_SUPER_MAGIC {
+                        return Ok(CgroupSetup::Hybrid);
+                    }
+                }
 
-    if cgroup2_mount.is_some() {
-        cgroups.push(Cgroup::V2);
+                return Ok(CgroupSetup::Legacy);
+            }
+        }
+        false => bail!("non default cgroup root not supported"),
     }
 
-    Ok(cgroups)
+    bail!("failed to detect cgroup setup");
```

Comment:
I don't think I'll actually reach the code here.

Reply:
It will if the filesystem type is not tmpfs.